### PR TITLE
fix issues with incremental compiler support

### LINKED ISSUE #45302: Fallback to release/profile correctly when experimental web compiler is enabled
Otherwise we always build a debug app in `flutter run`.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Currently the incremental compiler support does not handle release/profile mode. Fall back to the dwds runner, since we don't use the incremental compiler in that mode anyway.

Ensure we can handle the lack of a connected chrome instance for web-server device.

Fixes https://github.com/flutter/flutter/issues/45302

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/resident_web_runner_test.dart [hunk 1]
```diff
@@ -429,6 +431,49 @@ void main() {
...
+    unawaited(residentWebRunner.run(
+      connectionInfoCompleter: connectionInfoCompleter,
+    ));
+    await connectionInfoCompleter.future;
+    final OperationResult result = await residentWebRunner.restart(fullRestart: true);
+
+    expect(testLogger.statusText, contains('Restarted application in'));
+    expect(result.code, 0);
+    verify(mockResidentCompiler.accept()).called(2);
+	  // ensure that analytics are sent.
```
Comment (yjbanov) [inline packages/flutter_tools/test/general.shard/resident_web_runner_test.dart]: indent

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart [hunk 2]
```diff
@@ -448,10 +448,10 @@ class _ExperimentalResidentWebRunner extends ResidentWebRunner {
 
     try {
       if (fullRestart) {
-        await _wipConnection.sendCommand('Page.reload');
+        await _wipConnection?.sendCommand('Page.reload');
       } else {
-        await _wipConnection.debugger
-            .sendCommand('Runtime.evaluate', params: <String, Object>{
+        await _wipConnection?.debugger
+            ?.sendCommand('Runtime.evaluate', params: <String, Object>{
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart]: I think that change happens with NNBD, I get an analysis warning if I do foo?.bar.baz

File: packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart [hunk 3]
```diff
@@ -523,12 +523,14 @@ class _ExperimentalResidentWebRunner extends ResidentWebRunner {
...
-        await chrome.chromeConnection.getTab((ChromeTab chromeTab) {
-      return chromeTab.url.contains(debuggingOptions.hostname);
-    });
-    _wipConnection = await chromeTab.connect();
+    if (device.device is ChromeDevice) {
+      final Chrome chrome = await ChromeLauncher.connectedInstance;
+      final ChromeTab chromeTab =
+          await chrome.chromeConnection.getTab((ChromeTab chromeTab) {
+        return chromeTab.url.contains(debuggingOptions.hostname);
+      });
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart]: It was me! I tried to clean it up

File: packages/flutter_tools/test/general.shard/resident_web_runner_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/resident_web_runner_test.dart]: Fixed

**Pull request merged**